### PR TITLE
EN-3463 - service fail open implementation

### PR DESCRIPTION
**Pull request by pxaviad**

### CODE REVIEW

File: src/mod_perimeterx.c [hunk 1]
```diff
@@ -130,6 +137,45 @@ int px_handle_request(request_rec *r, px_config *conf) {
...
+    INFO(hc->server, "starting health_check thread");
+
+    const char *health_check_url = apr_psprintf(hc->server->process->pool, "%s/api/v1/kpi/status", hc->config->base_url);
+    CURL *curl = curl_easy_init();
+    CURLcode res;
+
+    long status_code;
+    while (true) {
+        status_code = 0;
+        apr_thread_mutex_lock(hc->config->px_errors_count_mutex);
```
Comment (barakpx) [inline src/mod_perimeterx.c]: px_errors_count_mutex need a better name, like health_check_cond_mutex

File: src/mod_perimeterx.c [hunk 2]
```diff
@@ -130,6 +137,45 @@ int px_handle_request(request_rec *r, px_config *conf) {
...
+    health_check_data *hc = (health_check_data*) data;
+    px_config *conf = hc->config;
+    INFO(hc->server, "starting health_check thread");
+
+    const char *health_check_url = apr_psprintf(hc->server->process->pool, "%s/api/v1/kpi/status", hc->config->base_url);
+    CURL *curl = curl_easy_init();
+    CURLcode res;
+
+    long status_code;
+    while (true) {
```
Comment (barakpx) [inline src/mod_perimeterx.c]: The health check thread loop should look like:
```
while (1) {
    // wait for condition and reset errors count on internal
    apr_thread_mutex_lock(conf->health_check_cond_mutex);
    while (apr_atomic_read32(&conf->px_errors_count) < conf->px_errors_threshold) {
        if (apr_thread_cond_timedwait(conf->health_check_cond, conf->health_check_cond_mutex, conf->health_check_interval) == APR_TIMEUP) {
            apr_atomic_set32(&conf->px_errors_count, 0);
        }
    }
    apr_thread_mutex_unlock(conf->health_check_cond_mutex);

    // do health check until success
    CURLcode res = CURLE_AGAIN;
    while (res != CURLE_OK) {
        curl_easy_setopt(curl, CURLOPT_URL, health_check_url);
        curl_easy_setopt(curl, conf->api_timeout);
        res = curl_easy_perform(curl);
        if (res != CURLE_OK && res != CURLE_OPERATION_TIMEDOUT) {
            apr_sleep(1000); // TODO(barak): should be configured with nice default
        }
    }
    apr_atomic_set32(&conf->px_errors_count, 0);
}
```

we still need to handle how we notify the thread to exit, for better shutdown cleanup. don't want to have crashes when we stop / restart

File: src/mod_perimeterx.c [hunk 3]
```diff
@@ -195,6 +243,32 @@ static void px_hook_child_init(apr_pool_t *p, server_rec *s) {
         apr_pool_cleanup_register(s->process->pool, cfg->activity_queue, apr_pool_cleanup_null, destroy_activity_queue);
         apr_pool_cleanup_register(s->process->pool, cfg->activity_queue, apr_pool_cleanup_null, destroy_thread_pool);
     }
+
+    // setting up health_check thread is service monitor enabled
+    if (cfg->px_service_monitor) {
+
+        health_check_data *hc_data= (health_check_data*)apr_palloc(s->process->pool, sizeof(health_check_data));
```
Comment (barakpx) [inline src/mod_perimeterx.c]: code style - missing space

File: src/mod_perimeterx.c [hunk 4]
```diff
@@ -195,6 +243,32 @@ static void px_hook_child_init(apr_pool_t *p, server_rec *s) {
...
+        apr_atomic_set32(&cfg->px_errors_count, 0);
```
Comment (barakpx) [inline src/mod_perimeterx.c]: not sure we need this call here - we can just set 0

File: src/mod_perimeterx.c [hunk 5]
```diff
@@ -195,6 +243,32 @@ static void px_hook_child_init(apr_pool_t *p, server_rec *s) {
...
+        hc_data->server = s;
+        hc_data->config = cfg;
+        apr_status_t rv;
```
Comment (barakpx) [inline src/mod_perimeterx.c]: we can merge the two lines

Comment (pxaviad) [inline src/mod_perimeterx.c]: how?

**Pull request merged**